signac/contrib/project.py

```diff
@@ -1910,26 +1893,19 @@ class JobsCursor:
     project : :class:`~signac.Project`
         Project handle.
     filter : Mapping
-        A mapping of key-value pairs that all indexed job state points are
-        compared against (Default value = None).
-
-    Notes
-    -----
-    Iteration is performed by acquiring job ids from the project using
-    :meth:`Project._find_job_ids`. When no filter (``filter = None``) is
-    provided, that method can take a much faster execution path, so not passing
-    a filter (or passing ``None`` explicitly) to this constructor is strongly
-    recommended over passing an empty filter (``filter = {}``) when iterating
-    over the entire data space.
+        A mapping of key-value pairs used for the query (Default value = None).
 
     """
 
     _use_pandas_for_html_repr = True  # toggle use of pandas for html repr
 
-    def __init__(self, project, filter=None, doc_filter=None):
+    def __init__(self, project, filter=None, *args, **kwargs):
         self._project = project
         self._filter = filter
 
+        doc_filter = next(iter(args), None) or kwargs.pop("doc_filter", None)
```

Comment:
I'd suggest that we add a deprecation warning if the `doc_filter` argument is not None.

Reply:
I had some reason for not doing that the first time, potentially related to signac-flow, but it might not be an issue after all. I'll add it.

Reply:
Ah, I remember why I hesitated. This code path should not be reachable if users follow our guidance not to instantiate the class directly. We don't test `JobsCursor.__init__` explicitly at all in signac, and I would prefer to remove the `doc_filter` constructor argument entirely, because we treat the constructor as private. @vyasr @csadorf Do you have opinions on this?

Reply:
I guess it would be fine then. In this case I would potentially even go so far and to just disallow it completely.

Reply:
Current versions of signac-flow do not use this constructor, so we should be safe to change this behavior if we treat the constructor as an internal API. The only uses of the `JobsCursor` constructor are in `Project.find_jobs` and `JobsCursor.groupby`, which should be replaced by `Project.find_jobs` for consistency. I will change that in this PR.